edk2toolext/environment/plugin_manager.py

```diff
@@ -82,32 +82,35 @@ def _load(self, PluginDescriptor):
             PluginDescriptor(PluginDescriptor): the plugin descriptor
         """
         PluginDescriptor.Obj = None
-        PythonFileName = PluginDescriptor.descriptor["module"] + ".py"
-        PyModulePath = os.path.join(os.path.dirname(os.path.abspath(
-            PluginDescriptor.descriptor["descriptor_file"])), PythonFileName)
-        logging.debug("Loading Plugin from %s", PyModulePath)
-        try:
-            with open(PyModulePath, "r") as plugin_file:
-                _module = imp.load_module(
-                    "UefiBuild_Plugin_" + PluginDescriptor.descriptor["module"],
-                    plugin_file,
-                    PyModulePath,
-                    ("py", "r", imp.PY_SOURCE))
 
+        py_file_path = PluginDescriptor.descriptor["module"] + ".py"
+        py_module_path = os.path.join(os.path.dirname(os.path.abspath(
+            PluginDescriptor.descriptor["descriptor_file"])), py_file_path)
+        py_module_name = "UefiBuild_Plugin_" + PluginDescriptor.descriptor["module"]
+
+        logging.debug("Loading Plugin from %s", py_module_path)
+
+        try:
+            spec = importlib.util.spec_from_file_location(
+                py_module_name, py_module_path)
+            module = importlib.util.module_from_spec(spec)
+            sys.modules[py_module_name] = module
+            sys.path.append(os.path.dirname(py_module_path))
```

Comment:
is it required to add the folder containing the module to the python path?  
This will impact all future loads of modules, right?

Reply:
`sys.path.append(os.path.dirname(py_module_path))` is the point of contention.

I don't want to speak for Sean, but he was either asking: 

1. Is it necessary to append this path to PYTHONPATH, the os environment, not just to sys.path
or
 2. He thinks appending to sys.path also appends to PYTHONPATH and he was asking the consequences of doing an unchecked append and polluting PYTHONPATH with the same paths from running a build over and over.
 
 If the concern is 1... It is not necessary to also append to PYTHONPATH as the plugins are loaded on every build, so sys.path variable will be freshly generated and loaded with the module paths each build.

If the concern is 2... sys.path is freshly built on every build from PYTHONPATH, however updates to sys.path are not mirrored to PYTHONPATH, so we don't have to worry about the appends to sys.path polluting PYTHONPATH.

The only slight concern is that if you have multiple plugins in the same directory, you'll append that directory multiple times.

@spbrogan let me know if I'm off the mark.

Reply:
Right, the behavior of the `sys.path` list is described in its documentation:
https://docs.python.org/3/library/sys.html#sys.path

For this:

> The only slight concern is that if you have multiple plugins in the same directory, you'll append that directory multiple times.

It is not different than what plugins previously did - If they understood what was wrong, they would have to `sys.path.append()` their module path before their import. Same problem, two+ plugins in the same directory, would append the same path.

But I agree that it doesn't make sense to push the same path twice. I updated to only append if needed:

```python
            if py_module_dir not in sys.path:
                sys.path.append(py_module_dir)
```